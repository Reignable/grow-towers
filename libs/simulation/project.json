{
  "name": "simulation",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/simulation/src",
  "prefix": "grow-towers",
  "tags": [],
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/angular:ng-packagr-lite",
      "outputs": ["{workspaceRoot}/dist/{projectRoot}"],
      "options": {
        "project": "libs/slots/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/slots/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/slots/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/simulation/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/simulation/**/*.ts", "libs/simulation/**/*.html"]
      }
    }
  }
}
