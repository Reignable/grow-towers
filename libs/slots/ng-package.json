{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/slots",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
