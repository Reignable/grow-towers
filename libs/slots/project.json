{
  "name": "slots",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/slots/src",
  "prefix": "grow-towers",
  "tags": [],
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/angular:ng-packagr-lite",
      "outputs": ["{workspaceRoot}/dist/{projectRoot}"],
      "options": {
        "project": "libs/slots/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/slots/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/slots/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/slots/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/slots/**/*.ts", "libs/slots/**/*.html"]
      }
    }
  }
}
