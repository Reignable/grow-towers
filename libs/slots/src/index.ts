export * from './lib/slots/slots.component'
