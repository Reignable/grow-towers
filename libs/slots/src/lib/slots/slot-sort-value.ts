export enum SlotSortValue {
  NUMBER = 'number',
  GROWTH_JOB_NAME = 'growthTray.growthJob.name',
  GROWTH_JOB_PROGRESS = 'growthTray.growthJob.progressPercentage',
}
