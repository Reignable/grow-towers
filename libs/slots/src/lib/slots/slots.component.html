<div class="controls">
  <grow-towers-growth-job-filter [control]="growthJobNameFilter" />
  <grow-towers-slots-sort [control]="slotSort" />
</div>
<ng-container *ngIf="sortedSlots$ | async as sortedSlots">
  <p *ngIf="!sortedSlots.length; else slotList">No slots matching filter</p>
  <ng-template #slotList>
    <ul class="slot-list">
      <grow-towers-slot-card
        *ngFor="let slot of sortedSlots; trackBy: trackBySlotNumber"
        [growthJobName]="slot.growthTray?.growthJob?.name"
        [growthJobProgressPercentage]="slot.growthTray?.growthJob?.progressPercentage"
        [growthTrayIdentifier]="slot.growthTray?.identifier"
      />
    </ul>
  </ng-template>
</ng-container>
