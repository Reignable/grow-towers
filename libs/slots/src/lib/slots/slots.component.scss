:host, .controls, .slot-list {
  display: flex;
  gap: 1rem;
}

:host, .slot-list {
  flex-direction: column;
}

.slot-list {
  padding: 0;
  list-style: none;
}
