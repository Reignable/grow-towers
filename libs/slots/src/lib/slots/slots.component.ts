import { CommonModule } from '@angular/common'
import { Component, inject } from '@angular/core'
import { FormControl } from '@angular/forms'
import { RouterLink } from '@angular/router'
import { GrowthJobName, Slot, TowerService } from '@grow-towers/simulation'
import { towerNumberRouteParam$ } from '@grow-towers/utils'
import { orderBy } from 'lodash'
import { CardModule } from 'primeng/card'
import { combineLatest, map, startWith, switchMap } from 'rxjs'
import { SlotCardComponent } from '../slot-card/slot-card.component'
import { GrowthJobFilterComponent } from './growth-job-filter/growth-job-filter.component'
import { SlotsSortComponent } from './slots-sort/slots-sort.component'

const trackBySlotNumber = (index: number, slot: Slot): number => slot.number

@Component({
  selector: 'grow-towers-slots',
  standalone: true,
  imports: [
    CardModule,
    CommonModule,
    GrowthJobFilterComponent,
    RouterLink,
    SlotCardComponent,
    SlotsSortComponent,
  ],
  templateUrl: './slots.component.html',
  styleUrls: ['./slots.component.scss'],
})
export class SlotsComponent {
  trackBySlotNumber = trackBySlotNumber

  towerNumber$ = towerNumberRouteParam$()

  towerService = inject(TowerService)

  slots$ = this.towerNumber$.pipe(
    switchMap(towerNumber => this.towerService.getTower(towerNumber)),
    map(tower => tower?.slots),
  )

  growthJobNameFilter = new FormControl<GrowthJobName | null>(null)

  filteredSlots$ = combineLatest([
    this.slots$,
    this.growthJobNameFilter.valueChanges.pipe(startWith(null)),
  ]).pipe(
    map(([slots, growthJobNameFilter]) => {
      if (!growthJobNameFilter) return slots
      return slots?.filter(slot => slot.growthTray?.growthJob?.name === growthJobNameFilter)
    }),
  )

  slotSort = new FormControl<string>('number', { nonNullable: true })

  sortedSlots$ = combineLatest([
    this.filteredSlots$,
    this.slotSort.valueChanges.pipe(startWith('number')),
  ]).pipe(
    map(([slots, slotSort]) => {
      if (!slotSort) return slots
      return orderBy(slots, slotSort, slotSort.includes('Percentage') ? 'desc' : 'asc')
    }),
  )
}
