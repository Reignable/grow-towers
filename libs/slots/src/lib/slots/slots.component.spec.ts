import { StaticDataFactory, TowerService } from '@grow-towers/simulation'
import { render, screen, waitFor } from '@testing-library/angular'
import { of } from 'rxjs'
import { SlotsComponent } from './slots.component'

const towerData = new StaticDataFactory().build()
const tower = towerData[0]

const renderSlotsComponent = () =>
  render(SlotsComponent, {
    providers: [
      {
        provide: TowerService,
        useValue: { towerData$: of(towerData), getTower: () => of(tower) },
      },
    ],
  })

describe(SlotsComponent, () => {
  it('Displays all slots for the given tower', async () => {
    await renderSlotsComponent()
    for (const slot of tower.slots) {
      if (!slot.growthTray?.identifier) break
      expect(screen.getByText(slot.growthTray.identifier)).toBeInTheDocument()
    }
  })

  it('Can filter slots by grow job type', async () => {
    const { fixture } = await renderSlotsComponent()
    fixture.componentInstance.growthJobNameFilter.setValue('Basil')
    await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(3))
  })

  it.each`
    sortType      | sortValue
    ${'number'}   | ${'number'}
    ${'job type'} | ${'growthTray.growthJob.name'}
    ${'progress'} | ${'growthTray.growthJob.progressPercentage'}
  `('Can sort jobs by $sortType', async ({ sortValue }) => {
    const { detectChanges, fixture } = await renderSlotsComponent()
    fixture.componentInstance.slotSort.setValue(sortValue)
    detectChanges()
    expect(screen.getAllByRole('listitem').map(({ textContent }) => textContent)).toMatchSnapshot()
  })
})
