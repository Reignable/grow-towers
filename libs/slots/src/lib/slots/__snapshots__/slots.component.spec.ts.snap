// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotsComponent Can sort jobs by job type 1`] = `
[
  " GT A1  Basil 10%",
  " GT A5  Basil 99%",
  " GT A10  Basil 42%",
  " GT A7  Kale 24%",
  " GT A8  Lettuce 30%",
  " GT A2  Rocket 20%",
  " GT A4  Strawberry 100%",
  " GT A6  Strawberry 15%",
  " GT A3  Tomato 0%",
  " GT A9  Tomato 13%",
]
`;

exports[`SlotsComponent Can sort jobs by number 1`] = `
[
  " GT A1  Basil 10%",
  " GT A2  Rocket 20%",
  " GT A3  Tomato 0%",
  " GT A4  Strawberry 100%",
  " GT A5  Basil 99%",
  " GT A6  Strawberry 15%",
  " GT A7  Kale 24%",
  " GT A8  Lettuce 30%",
  " GT A9  Tomato 13%",
  " GT A10  Basil 42%",
]
`;

exports[`SlotsComponent Can sort jobs by progress 1`] = `
[
  " GT A4  Strawberry 100%",
  " GT A5  Basil 99%",
  " GT A10  Basil 42%",
  " GT A8  Lettuce 30%",
  " GT A7  Kale 24%",
  " GT A2  Rocket 20%",
  " GT A6  Strawberry 15%",
  " GT A9  Tomato 13%",
  " GT A1  Basil 10%",
  " GT A3  Tomato 0%",
]
`;
